src/Windows/Avalonia.Win32.Automation/Marshalling/SafeArrayRef.cs

```diff
@@ -0,0 +1,339 @@
+        }
+    }
+
+    [Flags]
+    internal enum ADVANCED_FEATURE_FLAGS : ushort
+    {
+        FADF_AUTO = 0x0001,
+        FADF_STATIC = 0x0002,
+        FADF_EMBEDDED = 0x0004,
+        FADF_FIXEDSIZE = 0x0010,
+        FADF_RECORD = 0x0020,
+        FADF_HAVEIID = 0x0040,
+        FADF_HAVEVARTYPE = 0x0080,
+        FADF_BSTR = 0x0100,
+        FADF_UNKNOWN = 0x0200,
+        FADF_DISPATCH = 0x0400,
+        FADF_VARIANT = 0x0800,
+        FADF_RESERVED = 0xF008,
+    }
+
+    public void Destroy()
+    {
+        if (_ptr != default)
+        {
+            SafeArrayDestroy(_ptr);
+        }
+    }
+
+    public static T[]? ToArray<T>(SafeArrayRef? safearray)
+    {
+        if (safearray is null) return null;
+
+        return AccessData(safearray.Value, static (data, length) =>
+        {
+            var array = new T[length];
+            if (typeof(T) == typeof(sbyte))
+                Marshal.Copy(data, (byte[])(object)array, 0, length);
+            else if (typeof(T) == typeof(short))
+                Marshal.Copy(data, (short[])(object)array, 0, length);
+            else if (typeof(T) == typeof(int))
+                Marshal.Copy(data, (int[])(object)array, 0, length);
+            else if (typeof(T) == typeof(long))
+                Marshal.Copy(data, (long[])(object)array, 0, length);
+            else if (typeof(T) == typeof(byte))
+                Marshal.Copy(data, (byte[])(object)array, 0, length);
+            else if (typeof(T) == typeof(ushort))
+                Marshal.Copy(data, (short[])(object)array, 0, length);
+            else if (typeof(T) == typeof(uint))
+                Marshal.Copy(data, (int[])(object)array, 0, length);
+            else if (typeof(T) == typeof(ulong))
+                Marshal.Copy(data, (long[])(object)array, 0, length);
+            else if (typeof(T) == typeof(float))
+                Marshal.Copy(data, (float[])(object)array, 0, length);
+            else if (typeof(T) == typeof(double))
+                Marshal.Copy(data, (double[])(object)array, 0, length);
+            else if (typeof(T) == typeof(nint))
+                Marshal.Copy(data, (nint[])(object)array, 0, length);
+            else if (typeof(T) == typeof(nuint))
+                Marshal.Copy(data, (nint[])(object)array, 0, length);
+            else if (typeof(T) == typeof(string))
+            {
+                var pointers = new IntPtr[length];
+                Marshal.Copy(data, pointers, 0, array.Length);
+                for (var i = 0; i < length; i++)
+                {
+                    array[i] = (T)(object)Marshal.PtrToStringBSTR(pointers[i]);
+                }
+            }
+            else if (typeof(T).IsInterface)
+            {
+                var pointers = new IntPtr[length];
+                Marshal.Copy(data, pointers, 0, array.Length);
+                for (int i = 0; i < pointers.Length; i++)
+                {
+                    if (ComWrappers.TryGetObject(pointers[i], out var instance))
```

Comment:
`TryGetObject` will only unwrap an existing managed object, so it won't work for COM objects we didn't create. We should probably use `ComInterfaceMarshaller.ConvertToManaged()` here instead.

Reply:
I'd be a bit concerned about RCW ownership and potential finalization in that case.

Reply:
I don't want to care about unmanaged and RCW ownership in these cases.
We don't have to support that, since we don't consume automation interfaces from unmanaged code - we provide our managed wrappers instead.

Additionally, there is only one usage of SafeArrayRef.ToArray in IRawElementProviderAdviseEvents AdviseEventAdded+AdviseEventRemoved, but these two methods operate with arrays of integers. So we are safe there.

Reply:
To be even safer, I will add a logging message to the else code branch, and specify `[GeneratedComInterface(Options = ComInterfaceOptions.ManagedObjectWrapper)]` as much as possible (meaning, no code will be generated for consuming unmanaged-sourced com object wrappers)